PINCache/PINOperationQueue.m

```diff
@@ -0,0 +1,217 @@
+}
+
+- (void)dealloc
+{
+  pthread_mutex_destroy(&_lock);
+}
+
+- (id <PINOperationReference>)nextOperationReference
+{
+  [self lock];
+    id <PINOperationReference> reference = [NSNumber numberWithUnsignedInteger:++_operationReferenceCount];
+  [self unlock];
+  return reference;
+}
+
+- (id <PINOperationReference>)addOperation:(dispatch_block_t)block withPriority:(PINOperationQueuePriority)priority
+{
+  id <PINOperationReference> reference = [self nextOperationReference];
+  
+  NSMutableOrderedSet *queue;
+  switch (priority) {
+    case PINOperationQueuePriorityLow:
+      queue = _lowPriorityOperations;
+      break;
+      
+    case PINOperationQueuePriorityDefault:
+      queue = _defaultPriorityOperations;
+      break;
+      
+    case PINOperationQueuePriorityHigh:
+      queue = _highPriorityOperations;
+  }
+  
+  PINOperation *operation = [PINOperation operationWithBlock:block reference:reference];
+  
+  [self lock];
+    [queue addObject:operation];
+    [_allOperations addObject:operation];
+  [self unlock];
+  
+  [self scheduleOperations:NO];
+  
+  return reference;
+}
+
+- (void)cancelOperation:(id <PINOperationReference>)operationReference
+{
+  [self lock];
+    [_canceledOperations addObject:operationReference];
+  [self unlock];
+}
+
+- (void)scheduleOperations:(BOOL)onlyCheckSerial
+{
+  [self lock];
+    //get next available operation in order, ignoring priority and run it on the serial queue
+    if (_serialQueueBusy == NO) {
+      PINOperation *operation = [_allOperations firstObject];
+      if (operation) {
+        _serialQueueBusy = YES;
+        [self removeOperation:operation];
+        dispatch_async(_serialQueue, ^{
+          operation.block();
+          [self lock];
+            _serialQueueBusy = NO;
+          [self unlock];
+          
+          //see if there are any other operations
+          [self scheduleOperations:YES];
+        });
+      }
+    }
+  [self unlock];
+  
+  if (onlyCheckSerial) {
+    return;
+  }
+  
+  dispatch_async(_semaphoreQueue, ^{
+      dispatch_semaphore_wait(_concurrentSemaphore, DISPATCH_TIME_FOREVER);
+      [self lock];
+        PINOperation *operation = [self nextOperationByPriority];
+        [self removeOperation:operation];
+      [self unlock];
+    
+      if (operation) {
+        dispatch_async(_concurrentQueue, ^{
+          operation.block();
+          dispatch_semaphore_signal(_concurrentSemaphore);
+        });
+      } else {
+        dispatch_semaphore_signal(_concurrentSemaphore);
+      }
+  });
+}
+
+//Call with lock held
+- (PINOperation *)nextOperationByPriority
+{
+  PINOperation *operation = [_highPriorityOperations firstObject];
+  if (operation) {
+    return operation;
+  }
+  operation = [_defaultPriorityOperations firstObject];
+  if (operation) {
+    return operation;
+  }
+  operation = [_lowPriorityOperations firstObject];
+  return operation;
+}
+
+//Call with lock held
+- (void)removeOperation:(PINOperation *)operation
```

Comment:
Should be modified to always acquire the objects lock.  It's safer and doesn't violate the principle of least surprise for the next person working on the code.